### PR TITLE
fix: resolve ts compilerOptions.paths with prefixes and suffixes

### LINKED ISSUE #12851: [BUG] ts compilerOptions.paths with * in middle of value are not being resolved properly
**Context:**
- Playwright Version: 1.19.2
- Operating System: Mac
- Node.js version: v16.14.0

**Description**

[Recently added](https://github.com/microsoft/playwright/issues/7121#issuecomment-1009140849) support of typescript `compilerOptions.paths` config property doesn't work for values with `*` char in the middle. 
Example:

```
"paths": {
  "@company/prefix-*": [
    "packages/prefix-*/src"
  ]
}
```

The config above with error:

```sh
$ playwright test

Running 0 test using 0 worker


Error: Cannot find module '<path-to-project>project/packages/prefix-*/src'
Require stack:
...
```

Current support of paths works fine for values like `foo/*` or `foo`. But not for cases like `foo-*/bar`. However `tsc` is able to handle cases like this one.

**Reproduction repo**

https://github.com/ivankoleda/playwright-ts-paths-bug-reproduction

### PR DESCRIPTION
**Pull request by ivankoleda**
Resolves #12851
In addition to handling prefixes/suffixes in `compilerOptions.paths` added the support for matching the longest prefix.

### CODE REVIEW

File: packages/playwright-test/src/transform.ts [hunk 1]
```diff
@@ -101,23 +101,55 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
+      // * If module name can be matches with multiple patterns then pattern with the longest prefix will be picked.
+      // https://github.com/microsoft/TypeScript/blob/f82d0cb3299c04093e3835bc7e29f5b40475f586/src/compiler/moduleNameResolver.ts#L1049
+      const [prefix, suffix] = key.split('*');
+      if (prefix) {
+        matches = matches && specifier.startsWith(prefix);
+        matchedPartOfSpecifier = matchedPartOfSpecifier.replace(prefix, '');
+      }
+      if (suffix) {
+        matches = matches && specifier.endsWith(suffix);
+        matchedPartOfSpecifier = matchedPartOfSpecifier.replace(suffix, '');
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:124]: I think following input yields a bug. Let's add a test for it.

```js
key === '*bar';
specifier === 'barfoobar';
// should be 
matchedPartOfSpecifier === 'barfoo'
// actual 
matchedPartOfSpecifier === 'foobar'
```

File: packages/playwright-test/src/transform.ts [hunk 2]
```diff
@@ -101,23 +101,55 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
       continue;
     for (const value of values) {
-      const valueHasStar = value[value.length - 1] === '*';
-      let candidate = valueHasStar ? value.substring(0, value.length - 1) : value;
-      if (valueHasStar && keyHasStar)
-        candidate += specifier.substring(key.length - 1);
+      const valueHasStar = value.includes('*');
+
+      let candidate: string = value;
+      if (valueHasStar) candidate = candidate.replace('*', matchedPartOfSpecifier);
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:137]: style: we put body on the next line.

File: packages/playwright-test/src/transform.ts [hunk 3]
```diff
@@ -101,23 +101,55 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
       candidate = path.resolve(tsconfig.absoluteBaseUrl, candidate.replace(/\//g, path.sep));
       for (const ext of ['', '.js', '.ts', '.mjs', '.cjs', '.jsx', '.tsx']) {
-        if (fs.existsSync(candidate + ext))
-          return candidate;
+        if (fs.existsSync(candidate + ext)) {
+          const [keyPrefix] = key.split('*');
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:141]: Don't split in the loop. Btw, you also have it computed once above.

File: packages/playwright-test/src/transform.ts [hunk 4]
```diff
@@ -101,23 +101,55 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
+          // This comparassion is `>`, not `>=` to not override matches with same prefix length.
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:142]: No need for this comment.

**Review by pavelfeldman — COMMENTED** [review]
Nice! Some comments below.

File: packages/playwright-test/src/transform.ts [hunk 5]
```diff
@@ -100,23 +100,53 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
+        // * <MatchedStar> denotes part of the module name between <prefix> and <suffix>.
+        // * If module name can be matches with multiple patterns then pattern with the longest prefix will be picked.
+        // https://github.com/microsoft/TypeScript/blob/f82d0cb3299c04093e3835bc7e29f5b40475f586/src/compiler/moduleNameResolver.ts#L1049
+        if (keyPrefix) {
+          matches = specifier.startsWith(keyPrefix);
+          matchedPartOfSpecifier = matchedPartOfSpecifier.replace(keyPrefix, '');
+        }
+        if (keySuffix && matches) {
+          matches = specifier.endsWith(keySuffix);
+          matchedPartOfSpecifier = matchedPartOfSpecifier.replace(new RegExp(`${keySuffix}$`), '');
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:122]: I believe that this now has a new bug:
```js
keySuffix === '[abc]';
specifier === 'foo/[abc]';
```

use `matchedPartOfSpecifier.substring` instead.

File: packages/playwright-test/src/transform.ts [hunk 6]
```diff
@@ -100,23 +100,53 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
+
+      const keyHasStar = key.includes('*');
+      const [keyPrefix, keySuffix] = key.split('*');
+      if (keyHasStar) {
+        // * If pattern contains '*' then to match pattern "<prefix>*<suffix>" module name must start with the <prefix> and end with <suffix>.
+        // * <MatchedStar> denotes part of the module name between <prefix> and <suffix>.
+        // * If module name can be matches with multiple patterns then pattern with the longest prefix will be picked.
+        // https://github.com/microsoft/TypeScript/blob/f82d0cb3299c04093e3835bc7e29f5b40475f586/src/compiler/moduleNameResolver.ts#L1049
+        if (keyPrefix) {
+          matches = specifier.startsWith(keyPrefix);
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:117]: you can continue right away if it does not match, that way you don't need `matches`:

```js
if (!specifier.startsWith(keyPrefix))
  continue;
```

File: packages/playwright-test/src/transform.ts [hunk 7]
```diff
@@ -100,23 +100,52 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
+        }
+        if (keySuffix) {
+          if (specifier.endsWith(keySuffix))
+            matchedPartOfSpecifier = matchedPartOfSpecifier.substring(0, matchedPartOfSpecifier.lastIndexOf(keySuffix));
+          else continue;
+        }
+      } else {
+        if (specifier === key)
+          matchedPartOfSpecifier = specifier;
+        else continue;
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:128]: style: ditto

File: packages/playwright-test/src/transform.ts [hunk 8]
```diff
@@ -100,23 +100,52 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
+        // https://github.com/microsoft/TypeScript/blob/f82d0cb3299c04093e3835bc7e29f5b40475f586/src/compiler/moduleNameResolver.ts#L1049
+        if (keyPrefix) {
+          if (specifier.startsWith(keyPrefix))
+            matchedPartOfSpecifier = matchedPartOfSpecifier.replace(keyPrefix, '');
+          else continue;
+        }
+        if (keySuffix) {
+          if (specifier.endsWith(keySuffix))
+            matchedPartOfSpecifier = matchedPartOfSpecifier.substring(0, matchedPartOfSpecifier.lastIndexOf(keySuffix));
+          else continue;
```
Comment (pavelfeldman) [inline packages/playwright-test/src/transform.ts:123]: style: ditto

**Review by pavelfeldman — COMMENTED** [review]
Looks good, a couple of style nits and it is good to go!

File: packages/playwright-test/src/transform.ts [hunk 9]
```diff
@@ -100,23 +100,52 @@ export function resolveHook(filename: string, specifier: string): string | undef
...
+      const [keyPrefix, keySuffix] = key.split('*');
+      if (key.includes('*')) {
+        // * If pattern contains '*' then to match pattern "<prefix>*<suffix>" module name must start with the <prefix> and end with <suffix>.
+        // * <MatchedStar> denotes part of the module name between <prefix> and <suffix>.
+        // * If module name can be matches with multiple patterns then pattern with the longest prefix will be picked.
+        // https://github.com/microsoft/TypeScript/blob/f82d0cb3299c04093e3835bc7e29f5b40475f586/src/compiler/moduleNameResolver.ts#L1049
+        if (keyPrefix) {
+          if (!specifier.startsWith(keyPrefix))
+            continue;
+          matchedPartOfSpecifier = matchedPartOfSpecifier.replace(keyPrefix, '');
```
Comment (ivankoleda) [inline packages/playwright-test/src/transform.ts:118]: Actually this `replace` can also be replaced with `substring` for the sake of speed, let me do it as well

**Review by pavelfeldman — APPROVED** [review]
Approved. Will wait for a substring and merge!

**Pull request merged**